File: test/util/ImmutableDataTest.java
package util;

import java.util.Date;
import java.util.List;
import java.util.Arrays;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class ImmutableDataTest {
	
	@Test
	public void testImmutability() {
		
		final int[] srcNums = new int[]{1, 2, 3};
		final Date srcDate = new Date();
		final int srcHours = srcDate.getHours();
		final List<String> srcStrings = Arrays.asList("a", "b", "c");
		
		final ImmutableData data = new ImmutableData(srcNums, srcStrings, srcDate);
		
		final int[] resNums = data.getNumbers();
		
		// modify a number entry
		srcNums[2] = 99;
		assertEquals(resNums[2], 3);
		
		// modify a string entry
		final List<String> resStrings = data.getStrings();
		resStrings.set(0, "zzz");
		assertEquals(srcStrings.get(0), "a");
		assertEquals(data.getStrings().get(0), "a");
		
		// modify the date
		final Date resDate = data.getDate();
		resDate.setHours(10);
		assertEquals(srcDate.getHours(), srcHours);
		
	}
}

File: public/javascripts/src/test/ImmutableData.test.js
import ImmutableData from '../util/ImmutableData'

it('ImmutableData should preserve values', () => {

	const srcNumbers = [1,2,3];
	const srcObj = {
		a: {
			x: 99,
			y: -123,
			z: {
				q: 'qstring'
			}
		},
		b: 'a string'
	}
	const srcDate = new Date();
	const srcHours = srcDate.getHours();

	const data = new ImmutableData(srcNumbers, srcObj, srcDate);

	// modify a numbers entry
	expect(() => {
		const resNumbers = data.numbers;
		resNumbers[2] = 99;
	}).toThrow();
	expect(srcNumbers[2]).toEqual(3);

	// modify the nested object
	expect(() => {
		const resObj = data.data;
		resObj.a.x = 300;
	}).toThrow();
	expect(srcObj.a.x).toEqual(99);

	// modify the nested object
	expect(() => {
		const resObj = data.data;
		resObj.a.z.q = 999
	}).toThrow();
	expect(srcObj.a.z.q).toEqual('qstring');

	const resDate = data.date;
	resDate.setHours(1);
	expect(srcHours).toEqual(srcDate.getHours());
})